### PR TITLE
Message safety

### PR DESCRIPTION
**Pull request by carlhoerberg**
put sp in unacked only when delivered to consumer and ack or unack basic_get msgs after they've been deliviered

### CODE REVIEW

**Comment by carlhoerberg** [comment]
Extracted from #304

File: src/avalanchemq/queue/queue.cr [hunk 1]
```diff
@@ -696,25 +697,27 @@ module AvalancheMQ
...
     end
 
-    def basic_get(no_ack, force = false) : Envelope?
-      return nil if !@state.running? && (@state.paused? && !force)
+    def basic_get(no_ack, force = false, &blk : Envelope -> Nil) : Bool
+      return false if !@state.running? && (@state.paused? && !force)
       @last_get_time = Time.monotonic
       @get_count += 1
       if env = get(no_ack)
+        yield env
```
Comment (carlhoerberg) [inline src/avalanchemq/queue/queue.cr:705]: this might raise, and then the msg have to be requeued

File: src/avalanchemq/queue/queue.cr [hunk 2]
```diff
@@ -359,12 +359,14 @@ module AvalancheMQ
 
     private def deliver_to_consumer(c)
       # @log.debug { "Getting a new message" }
-      if env = get(c.no_ack, c)
+      if env = get(c.no_ack)
```
Comment (carlhoerberg) [inline src/avalanchemq/queue/queue.cr:362]: rescue exceptions in this block and requeue env from `get`

**Review by carlhoerberg — COMMENTED** [review]
todo

**Review by jage — APPROVED** [review]
Changes look good, if it's possible to cook up some specs it would be great.

**Comment by carlhoerberg** [comment]
yeah.. i think that's tricky because it's it would require the TCP socket to reset exactly while the server is sending the message, not sure to go about that..

**Comment by carlhoerberg** [comment]
Found a way to test it!

**Pull request merged**